canary/S3ObjectTransport.cpp

```diff
@@ -0,0 +1,916 @@
+        }
+
+        connectionManagerOptions.ConnectionOptions.Bootstrap = &m_canaryApp.GetBootstrap();
+        connectionManagerOptions.MaxConnections = 5000;
+
+        std::shared_ptr<Http::HttpClientConnectionManager> connManager =
+            Http::HttpClientConnectionManager::NewClientConnectionManager(connectionManagerOptions, g_allocator);
+
+        m_connManagers.push_back(connManager);
+    }
+}
+
+void S3ObjectTransport::MakeSignedRequest(
+    const std::shared_ptr<Http::HttpRequest> &request,
+    const Http::HttpRequestOptions &requestOptions,
+    SignedRequestCallback callback)
+{
+    String region = m_canaryApp.GetOptions().region.c_str();
+
+    Auth::AwsSigningConfig signingConfig(g_allocator);
+    signingConfig.SetRegion(region);
+    signingConfig.SetCredentialsProvider(m_canaryApp.GetCredsProvider());
+    signingConfig.SetService("s3");
+    signingConfig.SetBodySigningType(Auth::BodySigningType::UnsignedPayload);
+    signingConfig.SetSigningTimepoint(DateTime::Now());
+    signingConfig.SetSigningAlgorithm(Auth::SigningAlgorithm::SigV4Header);
+
+    m_canaryApp.GetSigner()->SignRequest(
+        request,
+        signingConfig,
+        [this, requestOptions, callback](
+            const std::shared_ptr<Aws::Crt::Http::HttpRequest> &signedRequest, int signingErrorCode) {
+            if (signingErrorCode != AWS_ERROR_SUCCESS)
+            {
+                if (callback != nullptr)
+                {
+                    callback(nullptr, signingErrorCode);
+                }
+                return;
+            }
+
+            std::shared_ptr<Aws::Crt::Http::HttpClientConnectionManager> connManager = GetNextConnManager();
+
+            connManager->AcquireConnection([this, requestOptions, signedRequest, callback](
+                                               std::shared_ptr<Http::HttpClientConnection> conn, int connErrorCode) {
+                if ((conn == nullptr || !conn->IsOpen()) && connErrorCode == AWS_ERROR_SUCCESS)
+                {
+                    connErrorCode = AWS_ERROR_UNKNOWN;
+                }
+
+                if (connErrorCode == AWS_ERROR_SUCCESS)
+                {
+                    MakeSignedRequest_SendRequest(conn, requestOptions, signedRequest);
+                }
+
+                if (callback != nullptr)
+                {
+                    callback(conn, connErrorCode);
+                }
+            });
+        });
+}
+
+void S3ObjectTransport::MakeSignedRequest_SendRequest(
+    const std::shared_ptr<Http::HttpClientConnection> &conn,
+    const Http::HttpRequestOptions &requestOptions,
+    const std::shared_ptr<Http::HttpRequest> &signedRequest)
+{
+    AWS_FATAL_ASSERT(conn->IsOpen());
+
+    Http::HttpRequestOptions requestOptionsToSend = requestOptions;
+    requestOptionsToSend.request = signedRequest.get();
+
+    ++m_activeRequestsCount;
+
+    // NOTE: The captures of the connection and signed request is a work around to keep these shared
```

Comment:
Less a workaround, and more CORRECT USE OF SHARED_PTR :muscle:

Reply:
Ah, you're right.  I'll take out that part of the comment.